cellxgene_schema_cli/cellxgene_schema/atac_seq.py

```diff
@@ -273,57 +273,67 @@ def validate_anndata_with_fragment(parquet_file: str, anndata_file: str) -> list
 
 
 def validate_fragment_no_duplicate_rows(parquet_file: str) -> Optional[str]:
-    df = ddf.read_parquet(parquet_file)
-    if len(df.drop_duplicates()) != len(df):
-        return "Fragment file has duplicate rows."
+    logger.info("starting validate_fragment_no_duplicate_rows")
+    t = ibis.read_parquet(f"{parquet_file}/**", hive_partitioning=True)
+    rows_per_chromosome = t["chromosome"].value_counts().execute()
+    msg = ""
+    for chromosome, count in rows_per_chromosome.itertuples(index=False):
+        n_unique = t.filter(t["chromosome"] == chromosome).distinct().count().execute()
+        if n_unique != count:
+            # TODO: See if we can improve error message
+            msg = "Fragment file has duplicate rows."
+            # msg += f"Chromosome {chromosome} has {count} rows but only {n_unique} are unique\n"
+    if msg:
+        return msg.strip()  # remove trailing newline
 
 
 def validate_fragment_start_coordinate_greater_than_0(parquet_file: str) -> Optional[str]:
-    df = ddf.read_parquet(parquet_file, columns=["start coordinate"])
-    series = df["start coordinate"] > 0
-    if not series.all().compute():
+    logger.info("starting validate_fragment_start_coordinate_greater_than_0")
+    df = ibis.read_parquet(f"{parquet_file}/**", hive_partitioning=True)
+    if not (df["start coordinate"] > 0).all().execute():
         return "Start coordinate must be greater than 0."
 
 
 def validate_fragment_barcode_in_adata_index(parquet_file: str, anndata_file: str) -> Optional[str]:
-    df = ddf.read_parquet(parquet_file, columns=["barcode"])
+    logger.info("starting validate_fragment_barcode_in_adata_index")
+    df = ibis.read_parquet(f"{parquet_file}/**", hive_partitioning=True)
+    barcode = set(df.select("barcode").distinct().execute()["barcode"])
```

Comment:
this looks expensive. Is converting to a set necessary. If this runs after checking for duplicates then we don't need get distinct.

Reply:
I generally have less worries about this one because:

* We know there are far fewer distinct entries
* The value we are checking for distinct-ness is smaller than the whole row